### ISSUE TITLE
Document optional dependencies

### ISSUE THREAD

**Issue by TomAugspurger** [dask/dask#5261]
https://dev.pandas.io/install.html#optional-dependencies for inspiration.

Also should include

- whether it's included in a standard `conda install dask`
- whether there's an extras for it.

Ideally we would have a CI build with all the oldest optional deps we support.

xref https://github.com/dask/dask/pull/5209

**Comment by pmk21** [comment]
Could I take up this issue? And any ideas on how to proceed?

**Comment by quasiben** [comment]
@pmk21 that would be great!  I'd recommend looking directly at the setup.py file:
https://github.com/dask/dask/blob/master/setup.py

You can also show dependencies with `conda`:

```
❯ conda search dask=2.5.0 --info
Loading channels: done
dask 2.5.0 py_0
---------------
file name   : dask-2.5.0-py_0.tar.bz2
name        : dask
version     : 2.5.0
build       : py_0
build number: 0
size        : 12 KB
license     : BSD 3-Clause
subdir      : noarch
url         : https://repo.anaconda.com/pkgs/main/noarch/dask-2.5.0-py_0.tar.bz2
md5         : 2df2ebb838e0aa2a58e83e1769f5d301
timestamp   : 2019-09-30 13:44:46 UTC
dependencies:
  - bokeh >=1.0.0
  - cloudpickle >=0.2.1
  - cytoolz >=0.7.3
  - dask-core 2.5.0.*
  - distributed >=2.5.0
  - fsspec >=0.5.1
  - numpy >=1.13.0
  - pandas >=0.21.0
  - partd >=0.3.10
  - python >=3.5
  - toolz >=0.7.3
```

**Comment by pmk21** [comment]
> https://github.com/dask/dask/blob/master/setup.py

Sorry a little lost here. The `setup.py` file shows all the required dependencies right? I'm not able to find the optional dependencies. 

> You can also show dependencies with `conda`:
> 
> ```
> ❯ conda search dask=2.5.0 --info
> Loading channels: done
> dask 2.5.0 py_0
> ---------------
> file name   : dask-2.5.0-py_0.tar.bz2
> name        : dask
> version     : 2.5.0
> build       : py_0
> build number: 0
> size        : 12 KB
> license     : BSD 3-Clause
> subdir      : noarch
> url         : https://repo.anaconda.com/pkgs/main/noarch/dask-2.5.0-py_0.tar.bz2
> md5         : 2df2ebb838e0aa2a58e83e1769f5d301
> timestamp   : 2019-09-30 13:44:46 UTC
> dependencies:
>   - bokeh >=1.0.0
>   - cloudpickle >=0.2.1
>   - cytoolz >=0.7.3
>   - dask-core 2.5.0.*
>   - distributed >=2.5.0
>   - fsspec >=0.5.1
>   - numpy >=1.13.0
>   - pandas >=0.21.0
>   - partd >=0.3.10
>   - python >=3.5
>   - toolz >=0.7.3
> ```

The conda output also shows the same dependencies as present in the `setup.py` file.

However I found quite a few dependencies in the CI files which aren't present in the `setup.py` file.

```yaml
dependencies:
  # required dependencies
  - python=3.7.*
  - numpy=1.17.*
  - pandas=0.25.*
  # test dependencies
  - pytest
  - pytest-xdist
  - moto
  - fastparquet
  - h5py
  - pytables
  - zarr
  - tiledb-py
  - fsspec>=0.5.1
  - sqlalchemy
  - pyarrow
  # other -- IO
  - bcolz
  - blosc
  - boto3
  - botocore
  - bokeh
  - httpretty
  - chest
  - s3fs
  - cloudpickle
  - crick
  - cytoolz
  - distributed==2.0.1
  - graphviz
  - ipython
  - lz4
  - numba
  - partd
  - psutil
  - requests
  - scikit-image
  - scikit-learn
  - scipy
  - toolz
  - python-snappy
  - sparse
  - cachey
  - python-graphviz
  - pandas-datareader
  - pip
  - pip:
      - cityhash
      - mmh3
      - xxhash
```

And also what should be the exact format of showing optional dependencies? And should these dependencies be displayed in the install section of the docs?

**Comment by quasiben** [comment]
I think the `extra_requires` are additional dependencies:

https://github.com/dask/dask/blob/13ddde5d68b4a68e3f16c8b5d4cd4ea3f8da578e/setup.py#L10-L20

Core bits of Dask can be optionally install: 
```
pip install "dask[array]"       # Install requirements for dask array
pip install "dask[bag]"         # Install requirements for dask bag
pip install "dask[dataframe]"   # Install requirements for dask dataframe
```
See https://docs.dask.org/en/latest/install.html#pip

I don't think pip has an easy way to list the dependencies easily -- but dask on it's own, no array, no dataframe is only dependent on Python, I think.  What I was thinking was just listing the dependencies listed in `extra_requires` as optional dependencies.  @TomAugspurger may have been thinking anything which throws an `ImportError`.  This could, without too much trouble, be done with `git grep/grep`:

```
(dask-test) bzaitlen@dgx15:~/GitRepos/dask$ git grep -B 1 "ImportError" # -B 1 means print 1 line before match
dask/__init__.py-    from .delayed import delayed
dask/__init__.py:except ImportError:
--
dask/__init__.py-    from .base import visualize, compute, persist, optimize, is_dask_collection
dask/__init__.py:except ImportError:
--
dask/array/__init__.py-
dask/array/__init__.py:except ImportError as e:
--
dask/array/__init__.py-    )
dask/array/__init__.py:    raise ImportError(str(e) + "\n\n" + msg)
--
dask/array/core.py-    from cytoolz.curried import pluck
dask/array/core.py:except ImportError:
--
dask/array/fft.py-    import scipy
dask/array/fft.py-    import scipy.fftpack
dask/array/fft.py:except ImportError:
```

From the above you can see that dask has optional dependencies on cytoolz, scipy, etc.

As for the CI files, yes those will have a lot more packages listed as testing requires more infrastructure.  For example, `py.test` is only necessary for testing

**Comment by pmk21** [comment]
Thanks @quasiben. I will work on it.

**Comment by jrbourbeau** [comment]
Closed by #5456
